libminifi/src/FlowController.cpp

```diff
@@ -449,13 +451,9 @@ void FlowController::initializeC2() {
 
   if (!c2_initialized_) {
     configuration_->setAgentIdentifier(identifier_str);
-    state::StateManager::initialize();
-    std::shared_ptr<c2::C2Agent> agent = std::make_shared<c2::C2Agent>(std::dynamic_pointer_cast<FlowController>(shared_from_this()), std::dynamic_pointer_cast<FlowController>(shared_from_this()),
+    c2_agent_ = std::make_shared<c2::C2Agent>(std::dynamic_pointer_cast<FlowController>(shared_from_this()), std::dynamic_pointer_cast<FlowController>(shared_from_this()),
```

Comment:
This was and still is a `shared_ptr` cycle between `C2Agent` and `FlowController`, which is not a bad practice and the reason for many of our memory leaks.
Since we are refactoring this we should solve this as well.

Reply:
As discussed, we will take this in another PR as it requires some major changes that might break interfaces.

Reply:
The base class changes of `FlowController` also break API backwards compatibility. If we care about bw compat in this PR, we should revert that as well. If we don't, then we should probably go ahead with the above as well.

I was not part of the cited discussion, so I might be missing something.